File: LC/Binary Search/search by value/378. Kth Smallest Element in a Sorted Matrix/binarysearch.py
class Solution:
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:
        n=len(matrix)
        l=-10**9
        r=10**9

        def check(mid):
            i=n-1
            j=0
            count=0

            while i>=0 and j<n:
                if matrix[i][j]<=mid:
                    count+=i+1
                    j+=1
                else:
                    i-=1
            return count

File: LC/Binary Search/search by value/875. Koko Eating Bananas/binarysearch.cpp
class Solution {
public:
    
    int minEatingSpeed(vector<int>& piles, int h) {
        int l=1;
        int r=*max_element(piles.begin(),piles.end());
        while(l<r){
            int mid=(l+r)/2;
            if (check(mid,piles)<=h){
                r=mid;
            }
            else{
                l=mid+1;
            }
        }
        return l;
    }
    int check(int mid,vector<int>& piles){
        int temp=0;
        for(auto & num:piles){
            if(num%mid==0){
                temp+=num/mid;
            }
            else{
                temp+=(num/mid)+1;
            }
        }
        return temp;       
    } 
};
